File: 2020/01/1.go
// https://adventofcode.com/2020/day/1
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"sort"
	"strconv"
)

func main() {
	nums, err := readNumbers()
	if err != nil {
		log.Fatal(err)
	}

	sort.Ints(nums)

	n, x, _ := part1(nums, 2020)
	fmt.Printf("part 1: %d * %d = %d\n", n, x, n*x)

	n, x, z, _ := part2(nums, 2020)
	fmt.Printf("part 2: %d * %d * %d = %d\n", n, x, z, n*x*z)
}

func readNumbers() ([]int, error) {
	var nums []int
	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		n, err := strconv.Atoi(scanner.Text())
		if err != nil {
			return nil, err
		}
		nums = append(nums, n)
	}
	return nums, scanner.Err()
}

func part1(nums []int, target int) (int, int, bool) {
	j := -1
	k := -1
	for i, n := range nums {
		x := target - n
		if z := sort.SearchInts(nums, x); z < len(nums) && nums[z] == x {
			j = i
			k = z
			break
		}
	}
	if j == -1 || k == -1 {
		return -1, -1, false
	}
	return nums[j], nums[k], true
}

func part2(nums []int, target int) (int, int, int, bool) {
	for _, n := range nums {
		t := target - n
		if x, z, ok := part1(nums, t); ok {
			return n, x, z, true
		}
	}
	return -1, -1, -1, false
}

File: 2025/01/1.c
/* https://adventofcode.com/2025/day/1 */

#include <stdio.h>
#include <stdlib.h>

int main(void)
{
    int dial = 50;
    int part1 = 0;
    int part2 = 0;
    char s[256];
    while (fgets(s, sizeof(s), stdin)) {
        int n = atoi(s+1);
        for (int i = 0; i < n; i++) {
            dial += s[0] == 'L' ? -1 : +1;
            if (dial == -1)
                dial = 99;
            else if (dial == 100)
                dial = 0;

            if (dial == 0)
                part2++;
        }
        if (dial == 0)
            part1++;
    }

    printf("part1 %d\npart2 %d\n", part1, part2);
}
